aarch64 support

tsahee opened the pull request:
This pull request addresses two aarch64 issues:
* adding support for aarch64 arch-timer
* outline-atomics compilation flag

This pull request is released under BSD license

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/MariaDB/server?pullRequest=1620) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/MariaDB/server?pullRequest=1620) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/MariaDB/server?pullRequest=1620) it.</sub>

grooverdan commented:
Regarding  aarch64 architecture timer, being a ASM can I assume it will work under FreeBSD aarch64? So `defined(__GNUC__)` isn't required?

mysqlonarm commented:
Both look good to me.

grooverdan commented:
Looking back `__GNUC__` is defined with clang on FreeBSD so this looks right to me.

grooverdan commented:
Minor, but for completeness could you add a routine.cycles id to https://github.com/MariaDB/server/blob/10.5/mysys/my_rdtsc.c#L361?  Its only used by unittest/mysys/my_rdtsc-t.c. Could you include the unit test results for the cycles timer in the commit?

dr-m commented on include/my_rdtsc.h:
```diff
@@ -164,6 +164,12 @@ static inline ulonglong my_timer_cycles(void)
+    ulonglong result;
+    __asm __volatile("mrs	%0, CNTVCT_EL0" : "=&r" (result));
+    return result;
+  }
```
Is this instruction available on all ARMv8 implementations? [MDEV-23249](https://jira.mariadb.org/browse/MDEV-23249) claims that the FreeBSD executables of MariaDB 10.3 or later are triggering SIGILL on the Raspberry Pi 3. In my understanding, that ought to be due to `clang`’s `__builtin_readcyclecounter()`. What code would that function generate?

dr-m replied:
Is this register guaranteed to be nonzero? @mysqlonarm posted [an experiment on MDEV-23249](https://jira.mariadb.org/browse/MDEV-23249?focusedCommentId=160667&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-160667) that suggests that 0 is returned on his hardware. I posted a follow-up comment with some analysis on the impact of a constant return value. I think that some InnoDB performance benchmarks could be necessary.

mysqlonarm replied:
@dr-m 

approach using __builtin_readcyclecounter returns 0 but the cntvct_el0 return expected values.

bash$ cat timer.cc 
#include <iostream>
using namespace std;

int main()
{
  unsigned long long result;
  __asm __volatile("mrs %0, CNTVCT_EL0" : "=&r" (result));
  cout << result << endl;

  cout << __builtin_readcyclecounter() << endl;
}
bash$ clang++-8 timer.cc 
bash$ ./a.out 
829795946130603
0
bash$ ./a.out 
829796087735762
0
bash$ ./a.out 
829796206079031
0
bash$ ./a.out 
829796414853453
0

tsahee replied:
It seems like he was reading PMCCNTR_EL0 - which is the cycle counter that belongs to the PMU. PMU could be set to e.g. only measure cycles in some execution states. It seems like a clang bug to use that reg for builtin_readcyclecounter. I'll dig.

I'm using CNTVCT_EL0 which is part of the architecture-timer. Always on, always non-decreasing.

I actually got to this patch because of severe performance issues. Without it, I got the constant "return 0" which was returned by get_rnd_value as you explain, which made my system crush as it couldn't get locks. With it, mariadb works well under load.

dr-m replied:
I realize that `__builtin_readcyclecounter()` was emitting code to read `pmccntr_el0` and not `cntvct_el0`. The inline assembler code appears to work for @mysqlonarm.

tsahee replied:
The first #if in that row is  # if __has_builtin(__builtin_readcyclecounter), which means clang will always go to it's own (wrong) implementation. I could test for aarch64 first, and not test for GNUC. It will fix mariadb on clang but code will look a little worse. Should I go this route?
In parallel, I'll dig into clang's __builtin_readcyclecounter and see if it could be fixed.

dr-m replied:
`__GNUC__` is defined (as 4) on `clang` possibly from the beginning. Thus, it might be sufficient to disable that `__builtin_readcyclecounter` on ARM and fall back to the `asm` code. It would indeed be nice if the wrong implementation were corrected in `clang`.

CyberBotX replied:
To note, 10.3 is not affected by this, as it does not contain this code, which was introduced in 10.4.7.

I have combined this patch along with testing that the target is not aarch64 in the test for `__builtin_readcyclecounter`. A complete patch that combines them can be found in FreeBSD [PR #248160](https://bugs.freebsd.org/bugzilla/attachment.cgi?id=216668&action=diff). I've mentioned in the mentioned JIRA report that this patch allows MariaDB 10.5 to run on my Raspberry Pi 3. Since 10.4's code is identical for this file, it should fix things there too.

dr-m replied:
@CyberBotX, thank you for confirming that this will address [MDEV-23249](https://jira.mariadb.org/browse/MDEV-23249).

@tsahee, can you please apply the ideas and rebase this pull request to 10.4? Please also refer to MDEV-23249 in the commit message. In that way, I can apply this pull request in such a way that GitHub identifies it as merged, not rejected.

tsahee replied:
I created https://github.com/MariaDB/server/pull/1635 for 10.4. Please let me know if anything is missing.

tsahee commented:
> Minor, but for completeness could you add a routine.cycles id to https://github.com/MariaDB/server/blob/10.5/mysys/my_rdtsc.c#L361? Its only used by unittest/mysys/my_rdtsc-t.c. Could you include the unit test results for the cycles timer in the commit?

addressed in the new pull request: https://github.com/MariaDB/server/pull/1635
Was not sure where you wanted the results - I added them to commit message.
On AWS m6g I got:
frequency: 121830845
resolution: 1
overhead: 1

dr-m commented:
This was closed by #1635 in 10.4, which will be merged to 10.5 some time before the 10.5.5 release.
